### PR TITLE
feat(NODE-4719): Add SDAM Logging Spec

### PR DESCRIPTION
**Pull request by aditi-khare-mongoDB**
### Description
Added support for logging on SDAM Events, and SDAM spec tests.

#### What is changing?
[All SDAM Events (except serverDescriptionChangedEvent) have a corresponding log.](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring-logging-and-monitoring.rst#log-messages) Using the `emitAndLog` function, we can output their information properly when  logging is turned on.

##### Is there new documentation needed for these changes?
Not until logging is released (NODE-5672)

#### What is the motivation for this change?

To implement standardized logging for SDAM. 

### Double check the following

- [ ] Ran `npm run check:lint` script
- [ ] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [ ] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [ ] Changes are covered by tests
- [ ] New TODOs have a related JIRA ticket

### CODE REVIEW

File: src/sdam/monitor.ts [hunk 1]
```diff
@@ -99,12 +100,16 @@ export class Monitor extends TypedEventEmitter<MonitorEvents> {
   /** @internal */
   [kMonitorId]?: MonitorInterval;
   rttPinger?: RTTPinger;
+  /** @internal */
+  override mongoLogger: MongoLogger;
```
Comment (nbbeeken) [inline src/sdam/monitor.ts:104]: We should not need this override correct? it should be inherited from the parent class

File: src/sdam/monitor.ts [hunk 2]
```diff
@@ -153,6 +157,8 @@ export class Monitor extends TypedEventEmitter<MonitorEvents> {
     }
 
     this.connectOptions = Object.freeze(connectOptions);
+
+    this.mongoLogger = mongoLogger;
```
Comment (nbbeeken) [inline src/sdam/monitor.ts:161]: nit, can we put this at the top of the constructor, next to where we attach the server seems like a good spot. The lower end of this function deals with some option reconfiguring that isn't relevant to this step.

File: src/sdam/events.ts [hunk 3]
```diff
@@ -1,21 +1,58 @@
...
 
 /**
- * Emitted when server description changes, but does NOT include changes to the RTT.
+ * The base export class for all monitoring events published from server discovery and monitoring.
  * @public
  * @category Event
  */
-export class ServerDescriptionChangedEvent {
-  /** A unique identifier for the topology */
+export abstract class ServerDiscoveryAndMonitoringEvent {
```
Comment (nbbeeken) [inline src/sdam/events.ts:21]: I don't see the introduction of this class in the kickoff, may want to loop the team in on this change since it is public API.

File: src/sdam/events.ts [hunk 4]
```diff
@@ -129,14 +166,17 @@ export class TopologyClosedEvent {
...
   connectionId: string;
-  /** Is true when using the streaming protocol. */
+  /** Is true when using the streaming protocol */
   awaited: boolean;
+  /** @internal */
+  name = SERVER_HEARTBEAT_STARTED;
 
   /** @internal */
-  constructor(connectionId: string, awaited: boolean) {
+  constructor(connectionId: string, awaited: boolean, topologyId: number) {
```
Comment (nbbeeken) [inline src/sdam/events.ts:178]: For this and the other heartbeat events I would put topologyId as the first parameter in the constructor, it has a more natural hierarchy: topologyId > connectionId.

File: src/mongo_logger.ts [hunk 5]
```diff
@@ -329,15 +358,36 @@ function attachCommandFields(
 
 function attachConnectionFields(
   log: Record<string, any>,
-  connectionPoolEvent: ConnectionPoolMonitoringEvent
+  connectionOrSDAMEvent: ConnectionPoolMonitoringEvent | ServerOpeningEvent | ServerClosedEvent
```
Comment (nbbeeken) [inline src/mongo_logger.ts:361]:
```suggestion
  event: ConnectionPoolMonitoringEvent | ServerOpeningEvent | ServerClosedEvent
```

I imagine there's more events that may end up here? I think we can just simplify

File: src/mongo_logger.ts [hunk 6]
```diff
@@ -329,15 +358,36 @@ function attachCommandFields(
...
+function attachServerHeartbeatFields(
+  log: Record<string, any>,
+  serverHeartbeatEvent:
+    | ServerHeartbeatFailedEvent
+    | ServerHeartbeatStartedEvent
+    | ServerHeartbeatSucceededEvent
+) {
+  const { awaited, connectionId } = serverHeartbeatEvent;
+  log.awaited = awaited;
+  log.driverConnectionId = 1;
```
Comment (nbbeeken) [inline src/mongo_logger.ts:384]: should this have been assigned to connectionId?

File: src/mongo_types.ts [hunk 7]
```diff
@@ -405,7 +405,7 @@ export class TypedEventEmitter<Events extends EventsDescription> extends EventEm
   /** @internal */
   protected component?: MongoLoggableComponent;
   /** @internal */
-  protected emitAndLog<EventKey extends keyof Events>(
+  emitAndLog<EventKey extends keyof Events>(
```
Comment (nbbeeken) [inline src/mongo_types.ts:408]: This shouldn't be necessary, we will discuss offline

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: src/mongo_logger.ts (hunk 6 above)
Comment (aditi-khare-mongoDB) [reply src/mongo_logger.ts:384]: Yes, It should have.. it's a remanent of personal debugging, I'll change it.

File: src/mongo_logger.ts (hunk 5 above)
Comment (aditi-khare-mongoDB) [reply src/mongo_logger.ts:361]: Sounds good, I've changed it locally, and will push one commit altogether.

File: src/sdam/monitor.ts [hunk 8]
```diff
@@ -99,12 +100,16 @@ export class Monitor extends TypedEventEmitter<MonitorEvents> {
...
+  override mongoLogger: MongoLogger;
+  /** @internal */
+  override component = MongoLoggableComponent.TOPOLOGY;
 
   get connection(): Connection | undefined {
     return this[kConnection];
   }
 
-  constructor(server: Server, options: MonitorOptions) {
+  constructor(server: Server, options: MonitorOptions, mongoLogger: MongoLogger) {
```
[the remark this answers is not in the document]
Comment (aditi-khare-mongoDB) [reply src/sdam/monitor.ts:112]: Bookkeeping: talked about this async

File: src/sdam/events.ts (hunk 3 above)
Comment (aditi-khare-mongoDB) [reply src/sdam/events.ts:21]: I see, I was sort of just following the precedent set by CMAP logging of having an abstract parent class that groups together a set of logging events. I'll ping the team to be sure though.

File: test/tools/unified-spec-runner/match.ts [hunk 9]
```diff
@@ -591,6 +633,30 @@ export function matchesEvents(
   }
 }
 
+export function filterLogs(
+  logsToIgnore: ExpectedLogMessage[],
+  actual: ExpectedLogMessage[],
+  entities: EntitiesMap
+): ExpectedLogMessage[] {
+  function isLogRelevant(log: ExpectedLogMessage) {
```
[the remark this answers is not in the document]
Comment (aditi-khare-mongoDB) [reply test/tools/unified-spec-runner/match.ts:641]: This is the filter function to filter out any logs that the spec tests list as "IgnoreMessages" so we don't fail the equality check.

**Review by nbbeeken — CHANGES_REQUESTED** [review]
We need to merge in main to remove the conflicts, Monitor has been changed. And we have the action item from slack to figure out if we can add topologyId to our heartbeat events.

File: src/mongo_types.ts [hunk 10]
```diff
@@ -410,7 +410,7 @@ export class TypedEventEmitter<Events extends EventsDescription> extends EventEm
     ...args: Parameters<Events[EventKey]>
   ): void {
     this.emit(event, ...args);
-    if (this.component) this.mongoLogger?.debug(this.component, args[0]);
+    if (this.component) this.mongoLogger?.debug?.(this.component, args[0]);
```
[the remark this answers is not in the document]
Comment (aditi-khare-mongoDB) [reply src/mongo_types.ts:413]: I made it an optional method to make the unit tests pass / I modified the mockServer constructor just now to bypass needing this.

File: test/spec/server-discovery-and-monitoring/unified/logging-replicaset.json [hunk 11]
```diff
@@ -352,9 +352,6 @@
                 "driverConnectionId": {
                   "$$exists": true
                 },
-                "serverConnectionId": {
```
[the remark this answers is not in the document]
Comment (aditi-khare-mongoDB) [reply test/spec/server-discovery-and-monitoring/unified/logging-replicaset.json]: I originally thought that the Node driver does not support serverConnectionId, since it is omitted in Node's cmap logging although it should be required.

[the remark this answers is not in the document]
Comment (aditi-khare-mongoDB) [reply test/spec/server-discovery-and-monitoring/unified/logging-replicaset.json]: I've added in serverConnectionId using handshake data. There's also a new item in the Q&A for the kickoff to this ticket concerning a missing TopologyDescriptionEvent emitted by our driver.

File: src/sdam/events.ts [hunk 12]
```diff
@@ -37,10 +49,12 @@ export class ServerDescriptionChangedEvent {
  * @category Event
  */
 export class ServerOpeningEvent {
-  /** A unique identifier for the topology */
+  /** A unique identifier for the topology  */
```
Comment (nbbeeken) [inline src/sdam/events.ts:52]: Here and a bunch of the comments like it we have an extra space, can we clean up the whitespace changes?

File: src/sdam/monitor.ts [hunk 13]
```diff
@@ -363,8 +374,10 @@ function checkServer(monitor: Monitor, callback: Callback<Document | null>) {
       }
 
       monitor.connection = conn;
-      monitor.emit(
+      monitor.emitAndLogHeartbeat(
         Server.SERVER_HEARTBEAT_SUCCEEDED,
+        monitor[kServer].topology.s.id,
+        conn.id,
```
Comment (nbbeeken) [inline src/sdam/monitor.ts:380]: We are passing in a mixture of driver-generated id and server connection id. This needs to also use the id that was provided on the handshake: `connection.hello.connectionId`

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: src/mongo_types.ts [hunk 14]
```diff
@@ -405,13 +405,26 @@ export class TypedEventEmitter<Events extends EventsDescription> extends EventEm
...
+  emitAndLogHeartbeat<EventKey extends keyof Events>(
+    event: EventKey | symbol,
+    topologyId: number,
+    serverConnectionId?: number | '<monitor>',
+    ...args: Parameters<Events[EventKey]>
+  ): void {
+    this.emit(event, ...args);
+    if (this.component) {
+      args[0].topologyId = topologyId;
+      args[0].serverConnectionId = serverConnectionId ?? null;
```
[the remark this answers is not in the document]
Comment (aditi-khare-mongoDB) [reply src/mongo_types.ts]: It likely does, I've changed it to use Object.assign to create a new object and also type annotated the new object.

File: src/sdam/events.ts (hunk 12 above)
Comment (aditi-khare-mongoDB) [reply src/sdam/events.ts]: Just cleaned up the extra spaces!

File: src/mongo_types.ts [hunk 15]
```diff
@@ -420,9 +426,14 @@ export class TypedEventEmitter<Events extends EventsDescription> extends EventEm
...
-      args[0].topologyId = topologyId;
-      args[0].serverConnectionId = serverConnectionId ?? null;
-      this.mongoLogger?.debug(this.component, args[0]);
+      const loggableHeartbeatEvent:
+        | LoggableServerHeartbeatFailedEvent
+        | LoggableServerHeartbeatSucceededEvent
+        | LoggableServerHeartbeatStartedEvent = Object.assign(
+        { topologyId: topologyId, serverConnectionId: serverConnectionId ?? null },
+        args[0]
+      );
```
Comment (nbbeeken) [inline src/mongo_types.ts:435]: Can we use spread syntax instead of Object.assign?

File: test/integration/server-discovery-and-monitoring/server_discover_and_monitoring.test.ts [hunk 16]
```diff
@@ -0,0 +1,10 @@
+import { loadSpecTests } from '../../spec';
+import { runUnifiedSuite } from '../../tools/unified-spec-runner/runner';
+
+describe('SDAM Unified Tests (Node Driver)', function () {
+  // TODO(NODE-5723): Remove this once the actual unified tests (test/spec/server-disovery-and-monitoring/logging) are passing
```
Comment (nbbeeken) [inline test/integration/server-discovery-and-monitoring/server_discover_and_monitoring.test.ts:5]: Can you summarize what is different about the tests from the specs here? Would it just be "logs for TopologyDescriptionChangedEvent are omitted"?

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: src/mongo_logger.ts [hunk 17]
```diff
@@ -576,7 +576,7 @@ function defaultLogTransform(
       log.message = 'Server heartbeat succeeded';
       log.durationMS = logObject.duration;
       log.serverConnectionId = logObject.serverConnectionId;
-      log.reply = stringifyWithMaxLen(logObject.reply, maxDocumentLength);
+      log.reply = EJSON.stringify(logObject.reply, { relaxed: true });
```
Comment (nbbeeken) [inline src/mongo_logger.ts:579]: What is the reason for this change?

Comment (aditi-khare-mongoDB) [reply src/mongo_logger.ts:579]: [According to the spec the reply field has to be relaxed](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring-logging-and-monitoring.rst#server-heartbeat-succeeded-log-message), which according to our [Node Logging Design Document](https://www.google.com/url?sa=t&rct=j&esrc=s&source=appssearch&uact=8&cd=0&cad=rja&q&sig2=HXozYfPozt3MApq7JZxyGQ&ved=0ahUKEwjDorKzr6OCAxUQwDcKHXsPBkE4ABABKAAwAA&url=https://drive.google.com/a/mongodb.com/open?id%3D1L3onvBqFaXuiN21sXMaFBRNS5kPnNYYckomt0s0ThmY&usg=AOvVaw0M0bd9q-Eh4qRu3zch2fhT), it will not be truncated.

Comment (aditi-khare-mongoDB) [reply src/mongo_logger.ts:579]: nvm, just extended maxDocLength in spec runner, ^^ above is not true.

File: test/tools/unified-spec-runner/match.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/tools/unified-spec-runner/match.ts:641]: Okay I followed up with the spec here, a few Qs:

- why check for level and component? shouldn't we be able to use resultCheck as a whole? (could be missing something)
  - What about the `failureIsRedacted` field?
- "which MUST be ignored on the corresponding client while executing [operations](https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.rst#test-operations)." This would imply that we need to invoke this whenever we go to collect a long, as opposed to filtering them at the end of the test run.

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: test/tools/unified-spec-runner/match.ts [hunk 18]
```diff
@@ -284,6 +289,8 @@ export function resultCheck(
   ) {
     // case to handle +0 and -0
     expect(Object.is(expected, actual)).to.be.true;
+  } else if (actual instanceof Int32 && typeof expected === 'number') {
+    expect(actual.value).to.equal(expected);
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/tools/unified-spec-runner/match.ts:293]: Okidoke, small change here, we try to avoid using `instanceof` to determine BSON types, so this should be: `actual._bsontype === 'Int32'`

File: test/tools/unified-spec-runner/match.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (W-A-James) [reply test/tools/unified-spec-runner/match.ts:641]: @nbbeeken re your second point, the sentence that follows clarifies that the filtering just has to happen before we evaluate the `messages` directive, which could be after all logs have been collected or filtered by the logger's `logDestination`.

Relevant sentence:
> The test runner MUST exclude all log messages from observed messages that match any of the messages in `ignoreMessages` array before `messages` evaluation.

@aditi-khare-mongoDB One other follow up question:
- what's the expected behaviour when there's overlap in the `messages` array and the `ignoreMessages` array?

**Review by W-A-James — CHANGES_REQUESTED** [review]

File: src/mongo_types.ts [hunk 19]
```diff
@@ -405,13 +411,32 @@ export class TypedEventEmitter<Events extends EventsDescription> extends EventEm
...
   /** @internal */
-  protected emitAndLog<EventKey extends keyof Events>(
+  emitAndLog<EventKey extends keyof Events>(
     event: EventKey | symbol,
     ...args: Parameters<Events[EventKey]>
   ): void {
     this.emit(event, ...args);
     if (this.component) this.mongoLogger?.debug(this.component, args[0]);
   }
+  emitAndLogHeartbeat<EventKey extends keyof Events>(
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/mongo_types.ts:421]: Good catch

[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/mongo_types.ts:421]: requesting changes for viz

**Review by nbbeeken — CHANGES_REQUESTED** [review]

**Review by W-A-James — APPROVED** [review]

**Review by nbbeeken — APPROVED** [review]
